rfcs/20210510-binary-size-reduction/README.md

```diff
@@ -0,0 +1,156 @@
+# oneDNN binary size reduction through build system.
+
+## Problem Statement.
+Several oneDNN users due to their usage model specifics would love to see their
+application linked statically with oneDNN (or shared object of the library
+itself) doesn't exceed specific disk space quota. One of such oneDNN users is
+OpenVINO toolkit. Our experiments showed that binary size of statically linked
+end-user application may take up to 64 MB [1]. Further analysis revealed several
+spots that contribute most:
+
+* The implementation of threading layer through templates. This is TBB specific,
+  OMP doesn't suffer from it. Takes ~14 MB.
+* The implementation of reorders through templates of plain C code and their
+  multiple instantiations. Takes ~9 MB.
+* The implementation of GEMM autogenerated JIT-ted kernels. Takes ~14 MB.
+* The implementation of convolution kernels apart of GEMM. Takes ~8 MB.
+* Rest functionality takes ~12 MB + ~2 MB of core API.
+
+## Background
+The reason why static linking can't solve the problem itself is the way library
+API is implemented. The call to `dnnl_primitive_desc_create` (which is also used
+underneath correspondent C++ API) relies on statically defined implementation
+lists which contain all primitives at ones. That means that if user calls the
+function (which is an essential part of the API), they get the full library in
+their application. And there are not many choices how to deal with such problem.
+One is to switch off implementations in compile time (this RFC proposal) or
+using a different API which would separate things on different levels (at least
+on primitive one).
+
+## Proposal
+To control the size, the proposal is to utilize the power of build system and
+let the user specify set of desired options. The idea is to have two sets of
+macros, one is external and would be defined by the build system based on user
+choices. This external set would control another set - internal - which lives
+inside the library and modifies implementation lists. Each primitive would get
+several macros to switch on/off certain implementations in compile time. Further
+optimizations of disk space will be provided by linker which shouldn't include
+unused code inside the end-user application unless certain scenario happens [2].
+
+As a part of this proposal, there are only two possible options to specify right
+now:
+* Propagation kind: inference or training (the default). Usage example, which
+  enables inference only:
+  - -DNNL_ENABLE_PROP_KIND=INFERENCE
+* Primitives. Usage example, which enables only convolution and matmul:
+  - -DDNNL_ENABLE_PRIMITIVES=CONVOLUTION;MATMUL
+
+This list may be extended in future upon request with data types and, possibly,
+ISA, if it's proven to help with further size reduction with minimal amount of
```

Comment:
I would recommend to consider conditional build by ISA in this RFC or at least in the nearest future. OV have multiple request to provide ability to build the product w/o SSE code path to decrease binary size. For client users it also might be useful to disable AMX primitives. All in all it should significantly impact library size.
Besides disabling particular primitives it should allow to remove some reorder implementetions from compilation. I understand that it assumes bigger effort from implementation standpoint, but the total gain will be valuable for sure.

Reply:
Hi @dmitry-gorokhov,
Thanks for this insight. I have several questions/comments:

> OV have multiple request to provide ability to build the product w/o SSE code path to decrease binary size. For client users it also might be useful to disable AMX primitives. All in all it should significantly impact library size.

This one sounds like if we have an ISA switch, then it should specify was ISA to exclude from list, right? Will it cover all or almost all requests for OpenVINO w.r.t. ISA support?

> All in all it should significantly impact library size.

Can't comment on this one since we haven't done any experiments. 

> Besides disabling particular primitives it should allow to remove some reorder implementations from compilation.

I believe we agreed to move most of them to JIT-ted implementation and leave a bare minimum working set for non-x64 builds. Do you think this won't be enough?
And I would stay away mixing ISA and formats devoted for certain ISA since this connection is very fragile and may change any minute, causing lots of issues with code promotions.

Reply:
> This one sounds like if we have an ISA switch, then it should specify was ISA to exclude from list, right? Will it cover all or almost all requests for OpenVINO w.r.t. ISA support?
Yes. Sounds good.

> I believe we agreed to move most of them to JIT-ted implementation and leave a bare minimum working set for non-x64 builds. Do you think this won't be enough?
I hope it will be enough. I understand your concerns regarding stability of such functionality. Lets leave it out of the scope of this RFC.